File: Medium/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.py
from collections import Counter
from typing import List


class Solution:
    def numTriplets(self, nums1: List[int], nums2: List[int]) -> int:
        def count_triplets(nums1, nums2):
            product_counts = Counter()
            count = 0

            for i in range(len(nums1)):
                for j in range(i + 1, len(nums1)):
                    product = nums1[i] * nums1[j]
                    product_counts[product] += 1

            for num in nums2:
                target = num * num
                count += product_counts[target]

            return count

        return count_triplets(nums1, nums2) + count_triplets(nums2, nums1)

File: Medium/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.ts
function numTriplets(nums1: number[], nums2: number[]): number {
    function countTriplets(nums1: number[], nums2: number[]): number {
        const productCounts = new Map<number, number>();
        let count = 0;

        for (let i = 0; i < nums1.length; i++) {
            for (let j = i + 1; j < nums1.length; j++) {
                const product = nums1[i] * nums1[j];
                productCounts.set(
                    product,
                    (productCounts.get(product) || 0) + 1
                );
            }
        }

        for (const num of nums2) {
            const target = num * num;
            count += productCounts.get(target) || 0;
        }

        return count;
    }

    return countTriplets(nums1, nums2) + countTriplets(nums2, nums1);
}
